tensorflow/core/util/sycl_util.h

```diff
@@ -0,0 +1,98 @@
+/* Copyright 2017 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#ifndef TENSORFLOW_USE_SYCL
+#error This file contains SYCL specific code and should only be included when \
+building with SYCL support.
+#endif
+
+#ifndef TENSORFLOW_CORE_UTIL_SYCL_UTIL_H_
+#define TENSORFLOW_CORE_UTIL_SYCL_UTIL_H_
+
+namespace tensorflow {
+
+// Need an atomic add for the MaxPoolGrad and MaxPool3DGrad kernels.
+//
+// For the device, this needs a pointer to global memory which isn't understood
+// by the host. The host should never be calling this method, but we provide
+// the header so that the host compiler can compile the functor.
+#ifdef __SYCL_DEVICE_ONLY__
+template <typename T>
+void SyclAtomicAdd(__attribute__((address_space(1))) T* address,
+                   const T increment);
+#else
+template <typename T>
+void SyclAtomicAdd(T* address, const T increment);
+#endif  // __SYCL_DEVICE_ONLY__
+
+#ifdef __SYCL_DEVICE_ONLY__
+// Use the OpenCL atomic uint operations to provide a floating point atomic add.
+// For the device we use the atomic compare-exchange builtin to keep trying to
+// add to the memory in a thread safe way. The union is needed as these
+// builtins are not availble for floating point types, only integer types, so
+// we do the addition on the float and the memory update on the uint.
+//
+// TODO(jwlawson): Remove once we have different type accessors for SYCL buffers
+// Providing a way to cast the types of buffers or accessors has been proposed
+// as a SYCL extension, so once this is available we can use an atomic
+// accessor and remove this.
```

Comment:
In particular, you can't turn those into atomic accessors because there is no atomic add for floats in OpenCL 1.2. Even OpenCL 2.0, which expanded the atomic operations available, doesn't have them.

In addition, the underlying buffer type is of int8_t, which has no atomic support either. Similarly, the cl::sycl::atomic type is not user-constructible, but even if it were, it would effectively be doing this anyway (or rather would call this SPIR-mangled function on device, and the C++11 version on host). That said, we'll take note of this feedback - seeing real-world usage is always useful.

Personally, I'm more of a fan of the less-abstract form of atomics where you simply operate atomically on addresses. though that's somewhat lower-level than SYCL aims for. I liked the atomic_view (later atomic_ref) proposals for C++, but I don't believe they made it into C++17.